mbzirc_ign/src/Wavefield.cc

```diff
@@ -0,0 +1,691 @@
+    this->amplitudes.clear();
+    this->phases.clear();
+    this->wavenumbers.clear();
+    this->steepnesses.clear();
+    this->directions.clear();
+
+    for (size_t i = 0; i < this->number; ++i)
+    {
+      const int n = i - this->number / 2;
+      const double scaleFactor = std::pow(this->scale, n);
+      const double a = scaleFactor * this->amplitude;
+      const double k = this->wavenumber / scaleFactor;
+      const double omega = this->DeepWaterDispersionToOmega(k);
+      const double phi = this->phase;
+      double q = 0.0;
+      if (a != 0)
+      {
+        q = std::min(1.0, this->steepness / (a * k * this->number));
+      }
+
+      this->amplitudes.push_back(a);
+      this->angularFrequencies.push_back(omega);
+      this->phases.push_back(phi);
+      this->steepnesses.push_back(q);
+      this->wavenumbers.push_back(k);
+
+      // Direction
+      const double c = std::cos(n * this->angle);
+      const double s = std::sin(n * this->angle);
+
+      const ignition::math::Vector2d d(
+        c * this->direction.X() - s * this->direction.Y(),
+        s * this->direction.X() + c * this->direction.Y());
+      directions.push_back(d);
+    }
+  }
+
+  // \brief Pierson-Moskowitz wave spectrum
+  public: double pm(double _omega, double _omegaP)
+  {
+    double alpha = 0.0081;
+    double g = 9.81;
+    return alpha * std::pow(g, 2.0) / std::pow(_omega, 5.0) * \
+      std::exp(-(5.0 / 4.0) * std::pow(_omegaP / _omega, 4.0));
+  }
+
+  /// \brief Recalculate for Pierson-Moskowitz spectrum sampling model
+  public: void RecalculatePms()
+  {
+    // Normalize direction
+    this->direction.Normalize();
+
+    // Derived mean values
+    this->angularFrequency = 2.0 * M_PI / this->period;
+    this->wavenumber = \
+      this->DeepWaterDispersionToWavenumber(this->angularFrequency);
+    this->wavelength = 2.0 * M_PI / this->wavenumber;
+
+    // Update components
+    this->angularFrequencies.clear();
+    this->amplitudes.clear();
+    this->phases.clear();
+    this->wavenumbers.clear();
+    this->steepnesses.clear();
+    this->directions.clear();
+
+    // Vector for spaceing
+    std::vector<double> omegaSpacing;
+    omegaSpacing.push_back(this->angularFrequency * (1.0 - 1.0 / this->scale));
+    omegaSpacing.push_back(this->angularFrequency * \
+                            (this->scale - 1.0 / this->scale) / 2.0);
+    omegaSpacing.push_back(this->angularFrequency * (this->scale - 1.0));
+
+    for (size_t i = 0; i < this->number; ++i)
+    {
+      const int n = i - 1;
+      const double scaleFactor = std::pow(this->scale, n);
+      const double omega = this->angularFrequency * scaleFactor;
+      const double pms = pm(omega, this->angularFrequency);
+      const double a = this->gain * std::sqrt(2.0 * pms * omegaSpacing[i]);
+      const double k = this->DeepWaterDispersionToWavenumber(omega);
+      const double phi = this->phase;
+      double q = 0.0;
+      if (a != 0)
```

Comment:
> nice, works for me. When testing in simple_demo.sdf world, I used a z pos value of -0.7 for the USV to be more stable otherwise there seems to be rocking movement.
> 
> Left some minor comments

I updated the wiki with the `-0.7` value.

Reply:
> use `math::equal` for floating point comparing?

Done in [47efdd9](https://github.com/osrf/mbzirc/pull/31/commits/47efdd99a68f9c014185c531313ea7d4c505d760)